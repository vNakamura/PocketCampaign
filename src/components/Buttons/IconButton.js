// @flow

import React from 'react';
import styled from 'styled-components';
import FaQuestion from 'react-icons/lib/fa/question';

import type { ReactChildren } from 'react-flow-types';

import Button from './Button';

const StyledButton = styled(Button)`
  flex: ${props => props.flex};
  transition: color .5s, border-color .5s;
`;
const Container = styled.div`
  display: flex;
  flex-direction: ${props => (props.textAtSide ? 'row' : 'column')};
  align-items: center;
  span {
    display: block;
    font-size: .8rem;
  }
  svg {
    font-size: ${props => (props.textAtSide ? '1rem' : '1.6rem')};
    margin: 4px;
  }
`;

type Props = {
  text?: string,
  active?: boolean,
  icon?: ReactChildren,
  flex?: number,
  onClick?: () => void,
  textAtSide?: boolean,
};

const IconButton = (props: Props) =>
  (<StyledButton flex={props.flex}><Container
    active={props.active}
    onClick={props.onClick}
    textAtSide={props.textAtSide}
  >
    {props.icon}
    {props.text ? <span>{props.text}</span> : null}
  </Container></StyledButton>);
IconButton.defaultProps = {
  icon: <FaQuestion />,
  flex: 0,
  active: false,
  onClick: undefined,
  text: undefined,
  textAtSide: false,
};

export default IconButton;
