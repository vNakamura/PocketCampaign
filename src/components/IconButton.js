import React from 'react';
import styled from 'styled-components';
import FaQuestion from 'react-icons/lib/fa/question';

import Button from 'Button';

const Container = styled(Button) `
  flex: 1;
  display: flex;
  flex-direction: column;
  align-items: center;
  span {
    display: block;
  }
  svg {
    font-size: 1.6rem;
    margin: 4px;
  }
  transition: color .5s, border-color .5s;
  border-bottom: solid 2px ${props => props.active ? props.theme.sidebar.itemHighlight : 'transparent'};
`;
export const IconButton = props => <Container active={props.active}>
  {props.icon? props.icon : <FaQuestion/>}
  <span>{props.text}</span>
</Container>

export default IconButton;
