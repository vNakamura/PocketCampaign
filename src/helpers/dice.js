// @flow

import React from 'react';
import droll from 'droll';
import seedrandom from 'seedrandom';
import replace from 'string-replace-to-array';

import NotationLink from '../components/Chat/NotationLink';

export const notationLimits = {
  minNumDice: 1,
  maxNumDice: 10,
  minNumSides: 2,
  maxNumSides: 100,
  minModifier: -1000,
  maxModifier: 1000,
};
export const diceNotationRegex = /\b([1-9]\d*)?d([1-9]\d*)([+-]\d+)?\b/gi;

type ResultError = {
  ok: false,
  message: string,
};
type ValidateResult = ResultError | { ok: true };
const validateRoll = (notation:string): ValidateResult => {
  const p = droll.parse(notation);
  if (!p) return { ok: false, message: 'Invalid dice notation' };
  if (p.numDice > notationLimits.maxNumDice) return { ok: false, message: 'Invalid dice notation. Too many dice!' };
  if (p.numSides < notationLimits.minNumDice) return { ok: false, message: 'Invalid dice notation. You need at least one die!' };
  if (p.numSides > notationLimits.maxNumSides) return { ok: false, message: 'Invalid dice notation. Too many sides!' };
  if (p.numSides < notationLimits.minNumSides) return { ok: false, message: 'Invalid dice notation. Too few sides!' };
  if (p.modifier > notationLimits.maxModifier) return { ok: false, message: 'Invalid dice notation. Modifier is too big!' };
  if (p.modifier < notationLimits.minModifier) return { ok: false, message: 'Invalid dice notation. Modifier is too big!' };
  return { ok: true };
};

export type ParseResult = ResultError | {
  ok: true,
  parsed: {
    numDice: number,
    numSides: number,
    modifier: number,
    minResult: number,
    maxResult: number,
    avgResult: number,
  }
};
export const parseRoll = (notation: string): ParseResult => {
  const v = validateRoll(notation);
  if (!v.ok) return v;
  const p = droll.parse(notation);
  return { ok: true, parsed: p };
};

export type RollResult = ResultError | {
  ok: true,
  rolls: number[],
  modifier: number,
  total: number,
};
export const roll = (notation: string, seed?: string): RollResult => {
  const v = validateRoll(notation);
  if (!v.ok) return v;

  if (seed !== undefined) seedrandom(seed, { global: true });
  const result = droll.roll(notation);
  return {
    ok: true,
    rolls: result.rolls,
    modifier: result.modifier,
    total: result.total,
  };
};

export const addLinksToText = (text: string) => {
  let i = 0;
  const result = replace(
    text,
    diceNotationRegex,
    (notation, numDice, numSides) => {
      i += 1;
      const v: ValidateResult = validateRoll(notation);
      if (v.ok && numSides > 1) return <NotationLink key={i} notation={notation} />;
      return notation;
    },
  );
  return result;
};
