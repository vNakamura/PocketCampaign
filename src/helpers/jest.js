import React from 'react';
import { ThemeProvider } from 'styled-components';
/* eslint-disable import/no-extraneous-dependencies */
import renderer from 'react-test-renderer';
import 'jest-styled-components';
/* eslint-enable import/no-extraneous-dependencies */
import theme from '../theme';

export const applyTheme = content => (
  <ThemeProvider theme={theme}>
    {content}
  </ThemeProvider>
);

export const renderWithTheme = content => renderer.create(
  applyTheme(content),
).toJSON();

export default {
  applyTheme,
  renderWithTheme,
};
