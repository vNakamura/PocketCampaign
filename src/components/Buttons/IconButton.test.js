import React from 'react';
import { mount } from 'enzyme';

import IconButton from './IconButton';
import { applyTheme, renderWithTheme } from '../../helpers/jest';

describe('components: Buttons/IconButton', () => {
  it('should render correctly', () => {
    const withoutText = renderWithTheme(<IconButton />);
    expect(withoutText).toMatchSnapshot();
    const withText = renderWithTheme(<IconButton text={'abcdef'} />);
    expect(withText).toMatchSnapshot();
  });

  it('shoud have a <span> with the text if the prop is set', () => {
    const withoutText = mount(applyTheme(<IconButton />));
    expect(withoutText.find('span')).toHaveLength(0);
    const withText = mount(applyTheme(<IconButton text={'asdfg'} />));
    expect(withText.find('span')).toHaveLength(1);
  });
});
