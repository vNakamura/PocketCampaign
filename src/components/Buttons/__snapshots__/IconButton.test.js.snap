// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components: Buttons/IconButton should render correctly 1`] = `
.c0 {
  -webkit-flex: 0;
  -ms-flex: 0;
  flex: 0;
  -webkit-transition: color .5s, border-color .5s;
  transition: color .5s, border-color .5s;
}

.c1 {
  background: none;
  border: none;
  cursor: pointer;
  color: #817b9e;
  font-family: Dosis;
  font-size: 1rem;
  padding: .2em .4em;
}

.c1:focus {
  outline: none;
}

.c1:hover,
.c1:focus {
  color: #e6e5f1;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c2 span {
  display: block;
  font-size: .8rem;
}

.c2 svg {
  font-size: 1.6rem;
  margin: 4px;
}

<button
  className="c0 c1"
  onClick={undefined}
  onMouseUp={[Function]}
>
  <div
    className="c2"
  >
    <svg
      fill="currentColor"
      height="1em"
      preserveAspectRatio="xMidYMid meet"
      style={
        Object {
          "color": undefined,
          "verticalAlign": "middle",
        }
      }
      viewBox="0 0 40 40"
      width="1em"
    >
      <g>
        <path
          d="m23.2 28v5.4q0 0.4-0.3 0.6t-0.6 0.3h-5.3q-0.4 0-0.7-0.3t-0.2-0.6v-5.4q0-0.3 0.2-0.6t0.7-0.3h5.3q0.4 0 0.6 0.3t0.3 0.6z m7.1-13.4q0 1.2-0.4 2.3t-0.8 1.7-1.2 1.3-1.3 1-1.3 0.8q-0.9 0.5-1.6 1.4t-0.6 1.5q0 0.4-0.2 0.8t-0.7 0.3h-5.3q-0.4 0-0.6-0.4t-0.2-0.8v-1q0-1.9 1.4-3.5t3.2-2.5q1.3-0.6 1.9-1.2t0.5-1.7q0-0.9-1-1.7t-2.4-0.7q-1.4 0-2.4 0.7-0.8 0.5-2.4 2.5-0.3 0.4-0.7 0.4-0.2 0-0.5-0.2l-3.7-2.8q-0.3-0.2-0.3-0.5t0.1-0.6q3.5-6 10.3-6 1.8 0 3.6 0.7t3.3 1.9 2.4 2.8 0.9 3.5z"
        />
      </g>
    </svg>
  </div>
</button>
`;

exports[`components: Buttons/IconButton should render correctly 2`] = `
.c0 {
  -webkit-flex: 0;
  -ms-flex: 0;
  flex: 0;
  -webkit-transition: color .5s, border-color .5s;
  transition: color .5s, border-color .5s;
}

.c1 {
  background: none;
  border: none;
  cursor: pointer;
  color: #817b9e;
  font-family: Dosis;
  font-size: 1rem;
  padding: .2em .4em;
}

.c1:focus {
  outline: none;
}

.c1:hover,
.c1:focus {
  color: #e6e5f1;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c2 span {
  display: block;
  font-size: .8rem;
}

.c2 svg {
  font-size: 1.6rem;
  margin: 4px;
}

<button
  className="c0 c1"
  onClick={undefined}
  onMouseUp={[Function]}
>
  <div
    className="c2"
  >
    <svg
      fill="currentColor"
      height="1em"
      preserveAspectRatio="xMidYMid meet"
      style={
        Object {
          "color": undefined,
          "verticalAlign": "middle",
        }
      }
      viewBox="0 0 40 40"
      width="1em"
    >
      <g>
        <path
          d="m23.2 28v5.4q0 0.4-0.3 0.6t-0.6 0.3h-5.3q-0.4 0-0.7-0.3t-0.2-0.6v-5.4q0-0.3 0.2-0.6t0.7-0.3h5.3q0.4 0 0.6 0.3t0.3 0.6z m7.1-13.4q0 1.2-0.4 2.3t-0.8 1.7-1.2 1.3-1.3 1-1.3 0.8q-0.9 0.5-1.6 1.4t-0.6 1.5q0 0.4-0.2 0.8t-0.7 0.3h-5.3q-0.4 0-0.6-0.4t-0.2-0.8v-1q0-1.9 1.4-3.5t3.2-2.5q1.3-0.6 1.9-1.2t0.5-1.7q0-0.9-1-1.7t-2.4-0.7q-1.4 0-2.4 0.7-0.8 0.5-2.4 2.5-0.3 0.4-0.7 0.4-0.2 0-0.5-0.2l-3.7-2.8q-0.3-0.2-0.3-0.5t0.1-0.6q3.5-6 10.3-6 1.8 0 3.6 0.7t3.3 1.9 2.4 2.8 0.9 3.5z"
        />
      </g>
    </svg>
    <span>
      abcdef
    </span>
  </div>
</button>
`;
